### DISCUSSION TITLE
连续发多个交易，最有最后一个交易链上能查到

### DISCUSSION THREAD

**Discussion by wwaayyaa** [nervosnetwork/ckb#2407] [General]
本地启动了ckb-cli server,然后通过http调用发款。
如图，我第一次连续发了3笔，只有最后一笔成功了。

![image](https://user-images.githubusercontent.com/6279538/77719282-e2d30a00-701f-11ea-8f78-7ab22b3996db.png)

![image](https://user-images.githubusercontent.com/6279538/77719177-9ee00500-701f-11ea-831c-956bd8cd9094.png)

然后又重新连发两笔交易，依然只有最后一笔交易成功了
![image](https://user-images.githubusercontent.com/6279538/77719343-0302c900-7020-11ea-9466-8c03c85699c3.png)
![image](https://user-images.githubusercontent.com/6279538/77719363-0d24c780-7020-11ea-874f-83cdf80c4717.png)

**Comment by doitian** [comment]
这些交易使用了相同的 inputs 吗？如果是的话只有一个会上链，其它交易会因为冲突被丢弃。如果是一个大的 cell 想连续发送交易有几种办法：

1. 先用一个 1 进多出的交易分成若干较小的 cells，然后使用小的 cells 分别转账
2. 直接用 1 进多出的交易向多个地址转账
3. 使用链式交易，第二个交易使用第一个交易的找零作为 inputs

**Comment by wwaayyaa** [comment]
> 这些交易使用了相同的 inputs 吗？如果是的话只有一个会上链，其它交易会因为冲突被丢弃。如果是一个大的 cell 想连续发送交易有几种办法：
> 
> 1. 先用一个 1 进多出的交易分成若干较小的 cells，然后使用小的 cells 分别转账
> 2. 直接用 1 进多出的交易向多个地址转账
> 3. 使用链式交易，第二个交易使用第一个交易的找零作为 inputs

我没有自己组装inputs、outputs，而是使用的如下图的方式进行多次发款。
![image](https://user-images.githubusercontent.com/6279538/77726773-b1643980-7033-11ea-8d56-67704de3ab80.png)

**Comment by doitian** [comment]
Assigned to @TheWaWaR

**Comment by TheWaWaR** [comment]
ckb-cli 如果要保证这种情况下三个交易都有效就会依赖当前 node 的交易池，感觉是一个需要讨论的设计。

建议发送完交易后查询该交易的状态变成 committed，才开始发送下一个交易。

**Comment by wwaayyaa** [comment]
看来只能先这样了。

**Comment by wwaayyaa** [comment]
@TheWaWaR 

ckb-cli server 下的rpc能否提供更多支持，目前只有几个方法。
```
transfer
get_capacity_by_address
get_capacity_by_lock_hash
get_live_cells_by_address
get_live_cells_by_lock_hash
get_live_cells_by_type_hash
get_live_cells_by_code_hash
```

所以像我这种使用server-rpc发款的用户，查询交易状态、是否被确认都没办法做到。只能再去浏览器。

**Comment by TheWaWaR** [comment]
可以直接通过向 ckb 节点发送 [`get_transaction`](https://github.com/nervosnetwork/ckb/tree/develop/rpc#get_transaction) rpc 请求来确认交易是否上链(status=commited)。

可以通过 `get_tip_block_number` rpc 拿到当前 tip 的高度，来计算那个交易已经被确认多少个块了。

**Comment by wwaayyaa** [comment]
今天又看了一下，我这边是通过http自动发款。之前已经实现了一笔一笔发。因为有时候交易会冲突然后就不见了。
所以希望能有1对多的例子可以参考。我看到 https://github.com/nervosnetwork/ckb-cli/wiki/Handle-Complex-Transaction 里面有说如何构建复杂的交易。但是是基于cli的。请问我应该去哪儿看到http相关的信息？包括获取livecell、拼装transaction（多个输出）、签名、发出这些完整的操作？

**Comment by doitian** [comment]
https://docs.nervos.org/docs/essays/rules
